compiler/rustc_mir/src/const_eval/eval_queries.rs

```diff
@@ -31,6 +31,13 @@ fn eval_body_using_ecx<'mir, 'tcx>(
 ) -> InterpResult<'tcx, MPlaceTy<'tcx>> {
     debug!("eval_body_using_ecx: {:?}, {:?}", cid, ecx.param_env);
     let tcx = *ecx.tcx;
+    assert!(
+        cid.promoted.is_some()
+            || matches!(
+                ecx.tcx.hir().body_const_context(cid.instance.def_id().expect_local()),
+                Some(ConstContext::Const | ConstContext::Static(_))
```

Comment:
Ah... I forgot about associated constants. We evaluate associated constants from other crates, because they may depend on generic parameters, so we can't eagerly evaluate them in the crate that defines them. Thus, I have an alternative for you that works across crates.

```suggestion
                ecx.tcx.def_kind(cid.instance.def_id()),
                DefKind::Const | DefKind::Static | DefKind::ConstParam | DefKind::AnonConst | DefKInd::AssocConst
```

I am not sure if that list should also contain `Ctor` or `Variant`, but you'll notice by the fact that the assertion is failing whether you need either.